Product Block Editor: Preprocessing for adding the channel visibility

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

To be compatible with Product Block Editor, it also needs to add the channel visibility block.

Since product attributes and channel visibility are different groupings in terms of code and UI presentation, it would be better to separate them into different sections in the Product Block Editor as well. Therefore, before adding the block, it requires some preprocessing:

- Move and rename `Admin\Product\Attributes\AttributesBlock` class to `Admin\ProductBlocksService` class.
   - This way, when the block corresponding to `ChannelVisibilityMetaBox` is added to it afterward, the code structure should make more sense.
- Change to create a dedicated group in the Product Block Editor for this extension.
- Add a new section to this extension's group in the Product Block Editor for channel visibility setup.
- Clean up the temporary codes that prevent unmapped inputs from being added to the Product Block Editor.

### Screenshots:

#### 📷 Simple product

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/6431f6ad-2ebe-4432-b2fb-272d818c5599)

#### 📷 Variable product

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/ffdf13d7-aeef-4a7e-8acd-f51442001bd7)

#### 📷 Variation product

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/5b2ae69e-1f19-48e4-b8f4-f8a945aa494f)

### Detailed test instructions:

#### 📌 Prepare test environment

1. Please refer to [the test preparation in PR 2151](https://github.com/woocommerce/google-listings-and-ads/pull/2151#prepare-test-environment) with **Woo 8.4.0**.
   - Currently, it can be tested with [8.4.0-rc.1](https://github.com/woocommerce/woocommerce/releases/tag/8.4.0-rc.1)
1. `npm install`
1. `npm start`

#### 📌 Test the dedicated group for GL&A

1. Go to edit a simple product, e.g. the **Beanie with Logo** imported from WC's sample products.
1. Check if there is a **Google Listings & Ads** tab, which is that dedicated group.
1. Switch to the **Google Listings & Ads** tab.
1. Check if there are **Channel visibility** and **Product attributes** sections.
   - Currently, the Channel visibility section is empty.
   - The Product attributes section should have attribute fields that correspond to the current product's type.
1. Repeat the above test steps with variable product and variation product types.

### Changelog entry

eason9487 committed:
Move and rename `Admin\Product\Attributes\AttributesBlock` class to `Admin\ProductBlocksService` class.

eason9487 committed:
Change to create a dedicated group in the Product Block Editor for this extension.

eason9487 committed:
Add a new section to this extension's group in the Product Block Editor for channel visibility setup.

eason9487 committed:
Clean up the temporary codes that prevent unmapped inputs from being added to the Product Block Editor.

eason9487 committed:
Add type references for variables related to the product block/template.

Address: https://github.com/woocommerce/google-listings-and-ads/pull/2151#discussion_r1424353087

puntope commented:
Not introduced in this PR, but since the Service is a Conditional service, what do you think of moving the conditionals from `register()` to the `is_needed()` function?

```php
public static function is_needed(): bool {
		return is_admin() && version_compare( WC_VERSION, '8.5', '>=' ) && $this->merchant_center->is_setup_complete();
	}
```

puntope commented on src/Admin/ProductBlocksService.php:
```diff
@@ -91,34 +93,55 @@ function () {
 		// https://github.com/woocommerce/woocommerce/blob/8.3.0/plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/AbstractProductFormTemplate.php#L16
 		$template_area = 'product-form';
 
-		// https://github.com/woocommerce/woocommerce/blob/8.3.0/plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/SimpleProductTemplate.php#L361
-		$block_id = 'product-attributes-section';
+		// https://github.com/woocommerce/woocommerce/blob/8.3.0/plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/SimpleProductTemplate.php#L18
+		// https://github.com/woocommerce/woocommerce/blob/8.3.0/plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/ProductVariationTemplate.php#L18
+		$block_id = 'general';
 
 		add_action(
 			"woocommerce_block_template_area_{$template_area}_after_add_block_{$block_id}",
```
Also not introduced in this PR but seems that we can just do:

`woocommerce_block_template_area_product-attributes-section_after_add_block_general`

eason9487 replied:
The use of PHP string interpolation here is to indicate that the action name is dynamically composited in Woo core.

puntope commented on src/Admin/ProductBlocksService.php:
```diff
@@ -91,34 +93,55 @@ function () {
+				);
+
+				/** @var Automattic\WooCommerce\Internal\Admin\Features\ProductBlockEditor\ProductTemplates\Section */
+				$channel_visibility_section = $group->add_section(
```
I think it should be `SectionInterface`, however, the variable `$channel_visibility_section` is not being used so we can omit it.

eason9487 replied:
Changed in https://github.com/woocommerce/google-listings-and-ads/pull/2179/commits/666cc9dcef530943e109981f5647af95e9a1a942.

> however, the variable `$channel_visibility_section` is not being used so we can omit it.

Sorry, the `$channel_visibility_section` will be used to add the channel visibility block later, and I should have added it in a subsequent PR but didn't notice. Since it was known the variable was going to be used, I didn't delete it.

puntope replied:
Thanks for the changes, in regards `Automattic\WooCommerce\Admin\Features\ProductBlockEditor\ProductTemplates\SectionInterface` which is used twice.

maybe we can do `use Automattic\WooCommerce\Admin\Features\ProductBlockEditor\ProductTemplates\SectionInterface;` 

and then only do `SectionInterface` without writing all the path.

eason9487 replied:
Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/2179/commits/34ca9e8c11055feb9f41f055376c51a69591d91f. I also adjusted the `add_product_attribute_blocks` method as the `add_block` method is defined in `SectionInterface` rather than `BlockInterface`.

puntope commented on src/Admin/ProductBlocksService.php:
```diff
@@ -91,34 +93,55 @@ function () {
-			function ( BlockInterface $images_section ) {
-				if ( ! $this->is_variation_template( $images_section ) ) {
+				// Please note that the simple and variable product types use the same product block template 'simple-product'.
+				if ( 'simple-product' !== $template->get_id() && ! $this->is_variation_template( $general_group ) ) {
```
💅  Maybe we can warp as well `'simple-product' !== $template->get_id()` into `$this->is_simple_template( $general_group )`

puntope commented on src/Admin/ProductBlocksService.php:
```diff
@@ -91,34 +93,55 @@ function () {
 
-				$section = $this->add_section( $images_section );
-				$this->add_blocks( $section );
+				/** @var Automattic\WooCommerce\Internal\Admin\Features\ProductBlockEditor\ProductTemplates\Group */
```
Shouldn't be `GroupInterface` instead of `Automattic\WooCommerce\Internal\Admin\Features\ProductBlockEditor\ProductTemplates\Group`

eason9487 replied:
Changed in https://github.com/woocommerce/google-listings-and-ads/pull/2179/commits/666cc9dcef530943e109981f5647af95e9a1a942.

puntope commented on src/Admin/ProductBlocksService.php:
```diff
@@ -91,34 +93,55 @@ function () {
+					]
+				);
+
+				/** @var Automattic\WooCommerce\Internal\Admin\Features\ProductBlockEditor\ProductTemplates\Section */
```
Shouldn't this be `SectionInterface`

puntope commented on src/Admin/ProductBlocksService.php:
```diff
@@ -91,34 +93,55 @@ function () {
+					]
+				);
+
+				/** @var Automattic\WooCommerce\Internal\Admin\Features\ProductBlockEditor\ProductTemplates\Section */
```
I think it should be `SectionInterface`

eason9487 replied:
Changed in https://github.com/woocommerce/google-listings-and-ads/pull/2179/commits/666cc9dcef530943e109981f5647af95e9a1a942.

puntope commented:
I see you added some hints for the IDE `@var BlockInterface` for example. 

They seem unneeded for me as knowing that is coming from BlocmInterface the IDE is smart enough to know whats type of these variables.

eason9487 committed:
Merge branch 'add/product-block-editor-custom-select-with-text-and-date-time' into tweak/product-block-editor-preprocessing-for-channel-visibility

eason9487 committed:
Document the `@var` tag in the `ProductBlocksService` class with the class of API definition rather than the actual class in runtime.

Address:
- https://github.com/woocommerce/google-listings-and-ads/pull/2179#discussion_r1436154516
- https://github.com/woocommerce/google-listings-and-ads/pull/2179#discussion_r1436154861
- https://github.com/woocommerce/google-listings-and-ads/pull/2179#discussion_r1436155273

eason9487 commented:
Thanks for the code review, @puntope! This PR is ready for the new round. Could you help with it?

> Not introduced in this PR, but since the Service is a Conditional service, what do you think of moving the conditionals from `register()` to the `is_needed()` function?
> 
> ```
> public static function is_needed(): bool {
> 		return is_admin() && version_compare( WC_VERSION, '8.5', '>=' ) && $this->merchant_center->is_setup_complete();
> 	}
> ```

The `is_needed` method in the `Conditional` interface is static, so the `$this->merchant_center` won't be available. Only moving the parts that can be run within the static method doesn't seem to help much. Therefore, I prefer to keep the current way.

> I see you added some hints for the IDE `@var BlockInterface` for example.
> 
> They seem unneeded for me as knowing that is coming from BlocmInterface the IDE is smart enough to know whats type of these variables.

It came from this code review suggestion: https://github.com/woocommerce/google-listings-and-ads/pull/2151#discussion_r1424353087.

puntope commented:
> The `is_needed` method in the `Conditional` interface is static, so the `$this->merchant_center` won't be available. Only moving the parts that can be run within the static method doesn't seem to help much. Therefore, I prefer to keep the current way.
> 
> >

You are right, didn't notice it was static.

puntope commented:
> It came from this code review suggestion: https://github.com/woocommerce/google-listings-and-ads/pull/2151#discussion_r1424353087.

At least for my ID is not necessary, bu ok to keep it if for others is.

puntope reviewed (approved):
LGTM ✅  Thanks for the changes @eason9487 

I left a couple of comments but approved in advance

eason9487 committed:
Refer to the `SectionInterface` class via `use` and adjust a related method's PHP doc.

Address: https://github.com/woocommerce/google-listings-and-ads/pull/2179#discussion_r1438181602
